qualtran/bloqs/comparison_gates.ipynb

```diff
@@ -0,0 +1,957 @@
+    "\n",
+    "When this decomposition is applied to the case where one input is a list of qubits and the the other is a classical number (i.e. a string of bits) this decomposition gives a $6n + \\mathcal{O}(1)$ T complexity since at the level right above the leafs half of the CSWAP operations degenerate to either identity or CNOTs, thus the number of CSWAPs becomes $\\approx \\frac{n}{2} + 2 \\times \\frac{n}{2} = \\frac{3n}{2}$, leadining to $4 \\times \\frac{3n}{2} = 6n$ T operations.\n",
+    "\n",
+    "However this is not the optimal way -in terms of T count- to compare a quantum number to a classical number which is $4n + \\mathcal{O}(1)$ T operations. The way to do can be found in appendix H of [Berry et al., 2019](\n",
+    "https://doi.org/10.48550/arXiv.1902.02134) and is based on reducing the problem to subtraction. The reduction in T count comes at the cost of linear depth as opposed to the logarithmic depth from the $6n$ decomposition.\n",
+    "\n",
+    "In what follows I will explain the T optimal ways for comparison in all cases."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "JxeFr2ClqHn9"
+   },
+   "source": [
+    "## Equality as a special case\n",
+    "Before we proceed to the comparison oracle we take a look at the equality oracle actually as a special case with a T complexity of only $4n + \\mathcal{O}(1)$, as it can be implemented as a qubit-wise And operation. This qubit wise And operation itself can be done using only $4n + \\mathcal{O}(1)$ as per [Babbush et al., 2018](https://doi.org/10.1103%2Fphysrevx.8.041015) and [Craig Gidney](https://algassert.com/post/1903).\n",
+    "\n",
+    "### Quantum-Classical Case\n",
+    "In the quantum classical case all we need to do is to compute qubit/bit wise equality\n",
+    "$$(q_0 = b_0) \\wedge \\cdots \\wedge (q_{n-1} = b_{n-1})$$\n",
+    "\n",
+    "In `cirq-ft` this is can be accomplished using the `And` gate as `And(cv=bits)` where `bits` are the bits of the classical number.\n",
+    "\n",
+    "### Quantum-Quantum Case\n",
+    "In the quantum-quantum case the equality is still bitwise $$(q_0 = p_0) \\wedge \\cdots \\wedge (q_{n-1} = p_{n-1})$$\n",
+    "\n",
+    "This is the same as the and $$(q_0 \\oplus p_0 = 0) \\wedge \\cdots \\wedge (q_{n-1}\\oplus p_{n-1} = 0)$$\n",
+    "\n",
+    "Where $\\oplus$ is the binary xor operation. This allows us to use `cirq_ft`'s `And` on the result of the xor."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing import Sequence\n",
+    "\n",
+    "try:\n",
+    "    import cirq\n",
+    "    import cirq_ft\n",
+    "except:\n",
+    "    !pip install --pre cirq\n",
+    "    !pip install --pre cirq-ft\n",
+    "    import cirq\n",
+    "    import cirq_ft"
```

Comment:
If you're planning on merging this to this repo, remove these installs since we assume the dependencies are installed in our notebooks. Otherwise, please don't use bare excepts; test for ImportError. The two pip installs can be combined into one. Let's not use `--pre` since I don't think we need to with the release of cirq[-ft] 1.2

Reply:
I'm planning to put it on the quantumai website .. not sure how that goes though